### PR TITLE
Migrate to rust-bitcoin 0.17

### PR DESCRIPTION
**Pull request by tamasblummer**
This is a minimal invasive migration, with the goal of compile and pass tests.
Preferred use of 0.17 features should follow once this is merged.

### CODE REVIEW

**Comment by tamasblummer** [comment]
@TheBlueMatt now also passes fuzz tests

File: src/util/migration.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+use bitcoin_hashes::hash160::Hash as Hash160;
+
+use std::io::Write;
+
+pub trait HashFromData {
+	fn from_data(buf: &[u8]) -> Self;
+}
+
+impl HashFromData for Sha256dHash {
+	fn from_data(buf: &[u8]) -> Sha256dHash {
```
Comment (TheBlueMatt) [inline src/util/migration.rs]: Hmm, we dont actually use from_data in that many places. Maybe just replace it with Hash::hash? https://docs.rs/bitcoin_hashes/0.3.1/bitcoin_hashes/trait.Hash.html#method.hash

File: src/util/transaction_utils.rs [hunk 2]
```diff
@@ -18,19 +19,27 @@ pub fn sort_outputs<T>(outputs: &mut Vec<(TxOut, T)>) {
...
-			Ordering::Less
-		} else if b.0.previous_output.txid.into_le() < a.0.previous_output.txid.into_le() {
-			Ordering::Greater
-		} else if a.0.previous_output.vout < b.0.previous_output.vout {
-			Ordering::Less
-		} else if b.0.previous_output.vout < a.0.previous_output.vout {
-			Ordering::Greater
-		} else {
-			Ordering::Equal
+		let c = cmp( &a.0.previous_output.txid, &b.0.previous_output.txid);
```
Comment (TheBlueMatt) [inline src/util/transaction_utils.rs]: nit can just do cmp(a, b).then(cmp(a, b))

Comment (tamasblummer) [reply src/util/transaction_utils.rs]: done

File: src/util/migration.rs (hunk 1 above)
Comment (tamasblummer) [reply src/util/migration.rs]: done

File: src/ln/msgs.rs [hunk 3]
```diff
@@ -782,6 +782,12 @@ impl From<::std::io::Error> for DecodeError {
 	}
 }
 
+impl From<::bitcoin_hashes::Error> for DecodeError {
```
Comment (TheBlueMatt) [inline src/ln/msgs.rs]: Hmm, I think you can skip this impl and just unwrap/expect in the impl Readable for Sha256dHash - the only Err return value is when you dont provide the right length value but we always do there.

**Comment by tamasblummer** [comment]
I really hate this squash before we merge policy. I accidentally squashed one commit more and now can not clean up since for some reason github does not accept the force push. I will create a new PR and close this.

**Comment by TheBlueMatt** [comment]
I dont really care if its completely squashed or not, but it is annoying to have intermediary commits which have failing tests (or worse, dont compile) as it can confuse git bisect and similar operations down the line.

**Comment by tamasblummer** [comment]
Those would not become part of master if you select "Sqash and Merge" while merging in the PR.

**Comment by TheBlueMatt** [comment]
Ah, right, yea, I suppose merge-commit-vs-just-commit is more of a personal preference thing. I'm always to happy to squash for folks.

**Comment by tamasblummer** [comment]
Not a problem for me too, but make mistakes and github behaves strangerly with force pushes.

**Pull request closed without merging**